docs/InferenceEngines.md

```diff
@@ -0,0 +1,521 @@
+
+(base) wyessen [~/dts/nn/xgmr/inference_engines/x-inference-engines/build/native/debug] $ od -f output.dat 
+0000000     5.850238e-01    9.525478e-01                                
+0000010
+
+(base) wyessen [~/dts/nn/xgmr/inference_engines/x-inference-engines/build/native/debug] $ 
+```
+
+#### Performance Monitoring (Linux Only)
+
+We can compile with the `-DENABLE_PERF_MONITORING` option (Linux only) to enable performance monitoring. After this, performance will be reported as in the following example. Note that not all VMs support this, and when running in a VM you may not get sensible results. Also, when running natively, it is often the case that you'll need to run with `sudo` to avoid permission errors (due to how performance metrics are gathered, elevated privileges may be required).
+
+```
+wyessen@raspberrypi:~/workbench/glow $ sudo ./bin/x-infer ./bundles/00_ffnn_nonquantized/ffnn.o ./bundles/00_ffnn_nonquantized/ffnn.weights -i ./data/00_ffnn_nonquantized/00_dummy_input.dat -l 2 -L 1 -m ffnn -n in -N save_out -o output.dat -p -t F32 -T F32
+
+Constant weights size       : 896 bytes
+Number of cases             : 1
+Number of CPU cycles (x1-e6): 0.014533
+```
+Note that CPU cycles are counted in millions.
+
+### Technical Notes and Notes on Dependencies
+
+1. Currently the engine is written in pure C (C99) and as such can be built and linked with pure C tools.
+2. As mentioned above, only UNIX-like platforms are currently supported (and for some features, only Linux). In addition,
+   GNU argument parsing library (GNU Argp) is required when building the stand-alone application. Please consult the appropriate documentation for your platform for help with installation. It is provided by default with the GNU build tools on many UNIX(-like) platforms, including Linux, and can be installed with `brew install argp-standalone` on OSX. When compiling on OSX, the additional `-DTARGET_OPTIONS=-largp` must be specified.
+3. Streaming is currently not supported, but this feature is planned.
+4. More than one named input, or more than one named output, is currently not supported. This support is planned in the future.
+5. Error reporting is rather rudimentary. More detailed error messages are needed. This is planned for the future.
+6. It may be possible to figure out input and output tensor lengths from the bundle, so that the `inlen` and `outlen` options can be safely dropped. This is planned for the future. Similarly for input and output types (although this would be more challenging).
+
+### A Note on the Initial Contribution
+
+The initial contribution of the `x-inference-engine` (as well as the corresponding documentation) was made as part of the open source contribution initiative by the XPERI Corporation (xperi.com). 
```

Comment:
Okay; a bit unusual, but allowable. Doesn't change the grant of copyright, or the Glow license in any way. I'd generally discourage these notations, just because they slow us down a bit to confirm wording w/legal, but if they want to make sure their name is in here somewhere, counsel says okay.

Reply:
> Okay; a bit unusual, but allowable. Doesn't change the grant of copyright, or the Glow license in any way. I'd generally discourage these notations, just because they slow us down a bit to confirm wording w/legal, but if they want to make sure their name is in here somewhere, counsel says okay.

@likethesky Understood. Our legal team is OK with us changing the headers to what Glow uses everywhere else. One question, though: are we allowed to identify ourselves somewhere following the source code legal headers? Something like the following, perhaps:

`/* Contributed by Xperi Corporation on [date] */`

So perhaps the entire header could look like this:

```
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Contributed by Xperi Corporation on [date] 
 */

It doesn't really matter where we identify ourselves, but it feels that somewhere in the header is the most natural place. Thanks!